### DISCUSSION TITLE
Pass Astro-generated components as an array to UI frameworks

### DISCUSSION THREAD

**Discussion by sHoskens** [withastro/roadmap#538] [Proposal]
# Summary

The ability to pass Astro-generated components as separate entities in an array to the props of a UI framework component (React).

# Background & Motivation

I noticed this "omission" from Astro while working on a simple carousel component. I needed one with some custom functionality and design, so started rolling it from scratch. My framework of choice for this was React. Now, when faced with how I was going to render the images, I preferred using Astro's build in, awesome `<Picture />` component. I know I can actually just replicate this functionality in React and be done with it, but for consistencies' sake (and keeping my codebase clean and DRY) I would have preferred to be able to use this. All the other images in my site are using it anyway.

My first attempt used slots, simply passing in the Picture elements. However, this turns all of the separate Picture elements into a single object with a single HTML string. (see [this issue](https://github.com/withastro/astro/issues/4926)) This is far from ideal when wanting to iterate over the separate children.

Named slots is an option which resolves this, but that would specifically require you to know the images in advance and hardcode them. Not pretty.

It looks to me like the ideal solution would be for Astro to allow me to pass in Astro components as separate elements in an array to a UI framework component, like so:

```
<MyReactCarousel client:visible images={[
  <Picture src="" alt="" />
  <Picture src="" alt="" />
  <Picture src="" alt="" />
]}>
```

# Goals

A **concise, bulleted-list** outlining the intended goals of this RFC. 

- Allow the passing of Astro components in an array to UI framework props.
- This could optionally be expanded by also allowing object syntax, for a dynamic alternative to named slots

**Comment by NathanCFoster** [comment]
After a little bit of digging, I found a workaround for this... its not pretty but it works

- At the top of your astro file paste these lines in
```js
let slots = [];
Object.keys(Astro.slots).forEach(element => slots.push(Astro.slots.render(element)));
let prop = await Promise.all(slots)
```
- Then pass that in to your react component as a prop like so `<ReactArr arr={prop} />`
- Finally set your html in your react component like so `<div dangerouslySetInnerHTML={{ __html: arr[0]}} />`

Hopefully this helps both you and any astro contributors for adding this.

**Reply by sHoskens** [reply]
Thanks for that! I think I'm gonna hold back from using that for now, since it makes what should be an easy to read piece of code quite complex and issue-prone. But it's good to know that there's at least some form of workaround in case I need it for more complex, dynamic cases.

**Comment by antonsarg** [comment]
Is there a better way now since I want to achieve the exact same thing? I want to pass Astro Image components to my React component which is using Swiper.

**Comment by manuhabitela** [comment]
Hey, I also stumbled on this. It would be super helpful to handle that usecase of array of elements as props or children in astro components.

**Comment by DevinWallKcl** [comment]
Render props and render prop like behavior would be wonderful to have.
